### PR TITLE
Provide test status to on_exit callbacks using minimal atoms.

### PR DESCRIPTION
**Pull request by christhekeele**
Spin-off from [discussion on main PR](https://github.com/elixir-lang/elixir/pull/15496#discussion_r3433840637)

### CODE REVIEW

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 1]
```diff
@@ -548,6 +548,24 @@ defmodule ExUnit.Runner do
...
+  defp run_exec_on_exit(%ExUnit.Test{state: nil}, pid, timeout) do
+    ExUnit.OnExitHandler.run(pid, :passed, timeout)
+  end
+
+  defp run_exec_on_exit(%ExUnit.Test{state: {:failed, _}}, pid, timeout) do
+    ExUnit.OnExitHandler.run(pid, :failed, timeout)
+  end
+
+  # Any other ExUnit.state outcome (test was excluded, skipped, invalid) and
+  # the runner shouldn't make it this far to ask the callback to be run anyways
```
Comment (christhekeele) [inline lib/ex_unit/lib/ex_unit/runner.ex:564]: I'm not actually 100% confident in this claim, but could not construct a scenario in the test suite where these edge cases actually invoked the callback, so at least fairly confident!

We could always pattern match apart the other `ExUnit.state()` outcomes here to be fool-proof (either calling `ExUnit.OnExitHandler.run` with the unreachable case or no-op returning `:ok`) but I did not want to clutter the implementation.

One benefit of the exhaustive approach (instead of the one I took here) is that if/when exhaustive type checking makes its way to the compiler, changes to the `ExUnit.state()` type would warn about unhandled/unreachable cases. Additionally, if the semantics of the runner were to change and attempt teardown during exclusion/skip/invalid, we'd be future-proof.

Thoughts?

Comment (christhekeele) [reply lib/ex_unit/lib/ex_unit/runner.ex:564]: I think I talked myself in to the more explicit version. :)
